### PR TITLE
Fix sliders being reversed incorrectly in the editor

### LINKED ISSUE #18042: Reversing slider in the editor doesn't work as expected
https://user-images.githubusercontent.com/78441708/166115757-5614bd5d-a285-4bcf-9238-08d221370eb2.mp4

map:
[Duskfox & Vixage - Aurora (Axilotl).zip](https://github.com/ppy/osu/files/8597492/Duskfox.Vixage.-.Aurora.Axilotl.zip)
@ 01:32:600

### Discussed in https://github.com/ppy/osu/discussions/18032

<div type='discussions-op-text'>

<sup>Originally posted by **Axilotl17** May  1, 2022</sup>
https://user-images.githubusercontent.com/78441708/166115757-5614bd5d-a285-4bcf-9238-08d221370eb2.mp4

it changes the shape drastically 

map:
[Duskfox & Vixage - Aurora (Axilotl).zip](https://github.com/ppy/osu/files/8597492/Duskfox.Vixage.-.Aurora.Axilotl.zip)
@ 01:32:600
</div>

### PR DESCRIPTION
**Pull request by Pasi4K5**
Fixes #18042

- [x] Depends on #24581

This updated `Reverse()` method now includes 2 new things:

### 1. Removing unnecessary `PathControlPoint`s at the end of a `SliderPath`
Let's use this slider as an example:
image
Previously, the control points at the end of the slider, which don't visibly affect the the path, were reverse just like all the other control points, which made the reversed slider look like this:
image
With the updated method, the control points at the end will be removed and the slider will be reversed as expected.
image
### 2. Recalculating the last perfect curve section of a slider
To fix the issue pointed out in #18042, `Reverse()` now takes the slider path of the last perfect curve and recalculates the middle control point to make sure it's always positioned between the perfect curve control point and the tail of the slider, so the slider shown in #18042 is now being reversed correctly:
image
By the way, stable reverses perfect curve sliders in a similar way. It doesn't ever remove any control points though.

Please note that this doesn't fix Bezier and Catmull sliders from being reversed incorrectly.
For example, this slider
image
when reversed, will still look like this
image
But, with the way things are now, I'm not sure if it's even possible to fix this due to how Bezier sliders work, but to be honest I don't think it's worth worrying about it since this doesn't majorly affect most sliders and mappers are used to this because the same issue exist in stable as well.

### CODE REVIEW

**Comment by bdach** [comment]
Before I even go into reviewing I would at least appreciate test cases for this. Should be simple enough to write, just check that the start and end positions of the sliders in the cited cases swap places rather than doing whatever else they used to do until now.

That said:

> Removing unnecessary `PathControlPoints` at the end of a `SliderPath`

This makes path reversal a destructive operation. I'm not sure how to feel about that.

**Comment by Pasi4K5** [comment]
> Before I even go into reviewing I would at least appreciate test cases for this. Should be simple enough to write, just check that the start and end positions of the sliders in the cited cases swap places rather than doing whatever else they used to do until now.

Will do soon.
> That said:
> 
> > Removing unnecessary `PathControlPoints` at the end of a `SliderPath`
> 
> This makes path reversal a destructive operation. I'm not sure how to feel about that.
> 
> Maybe if a reversal would truncate control points, it should prompt for confirmation or something. Not sure.

You can still use Ctrl+Z to restore the removed control points, so it's not really destructive, especially considering that with this change the shape of the slider stays the same when reversing it twice, which wasn't always the case previously.

**Comment by bdach** [comment]
> You can still use Ctrl+Z to restore the removed control points, so it's not really destructive

That doesn't make much sense given that we mark e.g.  slider-to-stream operation, or slider merge, as destructive. This change makes it kind of in-between, since there are no _objects_ being deleted, but some "data" is still lost (even if it has no value).

Maybe it's fine for it to be that way. Just a loose UX thought I guess.

**Comment by OliBomby** [comment]
I cant immediately come up with reasonable use cases where you'd want to reverse the control points exactly, so I guess in most cases it would be fine to corrupt the slider path a bit to keep the slider path the same. 
For the rare case where you'd want the current method of reverse it would be nice to have a workaround like being able to make the slider end fall exactly on the last control point so this new reverse method does the same as the old.

Also I think you should implement proper shortening of the bezier too. Its actually quite easy to do with Casteljau's algorithm. See the perfect curve to bezier converter.

**Comment by bdach** [comment]
> I cant immediately come up with reasonable use cases where you'd want to keep the extra bit of length, so I guess in most cases it would be fine to corrupt the control points a bit to keep the slider path the same.

Let's just keep as is for now then (make it discard the control points that don't do anything). In fact it could be a nice kind of conveyance for mappers that these points were bogus and not doing anything at all.

> Also I think you should implement proper shortening of the bezier too. Its actually quite easy to do with Casteljau's algorithm. See the perfect curve to bezier converter.

Oh interesting, if it's not prohibitively long LoC-wise then I'd like to see that done in this pull too.

**Comment by OliBomby** [comment]
Actually I take back what I said about the bezier. To find the right T value for a given length requires a more complicated algorithm so better to do that in a separate PR.

**Comment by Pasi4K5** [comment]
> For the rare case where you'd want the current method of reverse it would be nice to have a workaround like being able to make the slider end fall exactly on the last control point so this new reverse method does the same as the old.

Actually, this is *kind of* possible already. ***Kind of***. By changing the curve type of the last control point, the slider can be extended to the last point, ignoring the beat snap. You could then reverse the slider and resnap the sliderend. But that's a pretty hacky way to do it, and probably not intentional, so it might even get removed at some point.

> Actually I take back what I said about the bezier. To find the right T value for a given length requires a more complicated algorithm so better to do that in a separate PR.

I can *try* looking into that to see if I can come up with a solution for this myself, but IMO it's questionable if it's even worth giving it any attention in the near future, because, again, it doesn't significantly affect most sliders and it's been like this for years.

**Comment by OliBomby** [comment]
> I can *try* looking into that to see if I can come up with a solution for this myself, but IMO it's questionable if it's even worth giving it any attention in the near future, because, again, it doesn't significantly affect most sliders and it's been like this for years.

If you don't want to then I'll be happy to add the algorithm myself. I already have all the source code for the algorithm you need from the Mapping Tools repo.

**Comment by Pasi4K5** [comment]
> If you don't want to then I'll be happy to add the algorithm myself. I already have all the source code for the algorithm you need from the Mapping Tools repo.

I'm definitely not against it, so sure, go ahead if you want to.

File: osu.Game/Rulesets/Objects/SliderPathExtensions.cs
Comment (bdach) [inline osu.Game/Rulesets/Objects/SliderPathExtensions.cs:39]: This fix looks rather heavy-handed and it seems like it will create many more non-inherited control points than strictly necessary. Can you explain why this is necessary? Which cases are broken by not doing this? Is this fix covered by any of the added test cases?

File: osu.Game/Rulesets/Objects/SliderPathExtensions.cs
Comment (bdach) [inline osu.Game/Rulesets/Objects/SliderPathExtensions.cs:66]: Please use a normal for loop rather than do this.

File: osu.Game/Rulesets/Objects/SliderPathExtensions.cs
Comment (bdach) [inline osu.Game/Rulesets/Objects/SliderPathExtensions.cs:85]: Why is the arc approximated again if you've already got the midpoint from the old circle arc?

File: osu.Game/Rulesets/Objects/SliderPath.cs
Comment (bdach) [inline osu.Game/Rulesets/Objects/SliderPath.cs:205]: What is this clamp doing?

File: osu.Game.Rulesets.Osu.Tests/Editor/TestSceneSliderReversal.cs
Comment (bdach) [inline osu.Game.Rulesets.Osu.Tests/Editor/TestSceneSliderReversal.cs:47]: Nit: `.Prepend()` would have been much nicer here.

File: osu.Game.Rulesets.Osu.Tests/Editor/TestSceneSliderReversal.cs
Comment (bdach) [inline osu.Game.Rulesets.Osu.Tests/Editor/TestSceneSliderReversal.cs:99]: Nit: Please split assertions to one per step. Makes it much easier to debug potential failures.

File: osu.Game/Rulesets/Objects/SliderPathExtensions.cs
Comment (bdach) [reply osu.Game/Rulesets/Objects/SliderPathExtensions.cs]: Okay so first of all, if there is a way to produce a zero-length slider in the lazer editor, then that is a bug. We have been tightening that down so that it shouldn't be possible to create one.

Secondly the distinct thing has me baffled because it looks wrong and brittle. First of all, distinct is prone to failures like "the 2nd and the 5th control point out of 8 have the same position". Or like "this control point is 0.00001 to the left of the other". And finally, the code seems to be implicitly relying on the fact that `.Distinct()` doesn't reorder items, which is [not guaranteed](https://learn.microsoft.com/en-us/dotnet/api/system.linq.enumerable.distinct?view=net-7.0):

File: osu.Game/Rulesets/Objects/SliderPathExtensions.cs
Comment (bdach) [reply osu.Game/Rulesets/Objects/SliderPathExtensions.cs:38]: Okay, I see. I'd say two things as a follow-up to that:

- The "restoration" of control point types in the middle of the method confused me but I figured I'd leave that be at that point. It confused me mostly because it affected all of the points, but I assume it is done explicitly for this fix. If that is the case, I'd prefer that points which were temporarily modified to linear were collected to a list and then reverted from that list. (Or at least add an inline comment before restoring the control points that explains _why_ the control points are being reverted.)
- If possible I'd like to see test coverage that checks control point types too, to ensure the inherited linear points stay inherited at the end of it all. Up to you whether you wanna try, I may do this myself in the end.

File: osu.Game/Rulesets/Objects/SliderPath.cs
Comment (Pasi4K5) [reply osu.Game/Rulesets/Objects/SliderPath.cs:205]: I think we should ask @OliBomby about this. From what I've seen this method is primarily used for the stream creation tool (#21499).

File: osu.Game/Rulesets/Objects/SliderPathExtensions.cs
Comment (Pasi4K5) [reply osu.Game/Rulesets/Objects/SliderPathExtensions.cs]: > Okay so first of all, if there is a way to produce a zero-length slider in the lazer editor, then that is a bug. We have been tightening that down so that it shouldn't be possible to create one.

I submitted a bug report (#24559).
> So if the goal here is to check that the last N control points are sorta in the same place [...]

Just to make sure there is no misunderstanding: The goal of this line https://github.com/ppy/osu/blob/cd706734636cd3d8da5bef4e9078c4b4d70cd5f9/osu.Game/Rulesets/Objects/SliderPathExtensions.cs#L41 was to get the `segmentEnds` without the duplicates *at the end of the array*, not to *actually* remove control points which are in the same place as others. Since `GetSegmentEnds()` gives us the same number for all control points beyond the sliderend, this can be used to detect how many control points the reversed slider shall have.

 That being said, this
> "the 2nd and the 5th control point out of 8 have the same position"

and this
> the code seems to be implicitly relying on the fact that .Distinct() doesn't reorder items, which is [not guaranteed](https://learn.microsoft.com/en-us/dotnet/api/system.linq.enumerable.distinct?view=net-7.0)

is still valid, so I replaced `.Distinct()` with a custom method (`truncateEndingDuplicates()`) instead.

File: osu.Game/Rulesets/Objects/SliderPath.cs
Comment (OliBomby) [reply osu.Game/Rulesets/Objects/SliderPath.cs:205]: Hello, yeah the values returned by `GetSegmentEnds()` are kinda wierd.

The zero prepended to the output was intentional because it had the nice property of basically giving you the start and end of every segment, but in hindsight its kinda dumb and useless because it doesn't follow the theme of the name or documentation at all.
The dividing by the `calculatedLength` is an error. The values it returns are supposed to work as an argument to `PositionAt()`, and for that it's supposed to divide by `Distance`. When used for the stream creation tool, this error wasn't obvious because I never set an `ExpectedDistance` on the `SliderPath` for streams.

The `IndexOutOfRangeException` might be an issue in `calculateLength()` which doesn't consider removing elements from `segmentEnds` if the `expectedDistance` is sufficiently low to have fewer segments.

Comment (Pasi4K5) [reply osu.Game/Rulesets/Objects/SliderPath.cs]: Thanks for the response!
> The IndexOutOfRangeException might be an issue in calculateLength() which doesn't consider removing elements from segmentEnds if the expectedDistance is sufficiently low to have fewer segments.

The root cause of this can indeed be found in `calulateLength()`. 
https://github.com/ppy/osu/blob/58bffa13cd284891ebd735c216ca9bdab94ca8c8/osu.Game/Rulesets/Objects/SliderPath.cs#L317-L326 When shortening the path, the last element of `cumulativeLength` is removed until it matches the `expectedDistance`. Since `segmentEnds` is basically just a list of indexes for `cumulativeLength`, `segmentEnds[^1]` is decremented to make sure that it's not out of bounds (L325). The oversight here is that all elements *before* `segmentEnds[^1]` can still be out of bounds. In other words, only the last segment is shortened to the expected distance, while really all of them should be.

That's why the exception always occurs when a slider has more than 1 segment end after its tail.

My suggested solution would be to just cap all the `segmentEnds` after the path is shortened. I committed this as well as fixes for the other 2 issues you mentioned and adjusted my code accordingly.

Comment (OliBomby) [reply osu.Game/Rulesets/Objects/SliderPath.cs]: > By that do you mean that segment ends beyond the ExpectedDistance will give you a number > 1?

Yes. I think like that the method will have the most value. It will return one value for every segment in the slider path (unshortened), and the values will not be trivial values such as 0 or 1.

Comment (OliBomby) [reply osu.Game/Rulesets/Objects/SliderPath.cs]: @Pasi4K5 I've made a PR incorporating my fix https://github.com/ppy/osu/pull/24581

Comment (Pasi4K5) [reply osu.Game/Rulesets/Objects/SliderPath.cs]: > You will be able to easily find the segments which can be trimmed. If a segment starts before 1 and ends after 1 then it needs to be shortened. If a segment starts after 1 then it can be removed completely.

Perfect! That makes things a lot simpler. I'll wait until it got merged and then update my code.

Comment (OliBomby) [reply osu.Game/Rulesets/Objects/SliderPath.cs]: You could merge my branch into yours right now and use it immediately. Then you can merge this PR after mine is merged.

**Review by bdach — APPROVED** [review]

**Pull request merged**
